### PR TITLE
Fix Notion code exchange

### PR DESCRIPTION
**Pull request by mifopen**
Copied default implementation of ExchangeCodeAsync and added Basic authorization
https://developers.notion.com/docs/authorization#exchanging-the-grant-for-an-access-token

### CODE REVIEW

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 1]
```diff
@@ -26,6 +32,45 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
...
+                context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+            }
+
+            var requestContent = new FormUrlEncodedContent(tokenRequestParameters!);
+
+            using var requestMessage = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+            requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            byte[] byteArray = Encoding.ASCII.GetBytes(Options.ClientId + ":" + Options.ClientSecret);
+            requestMessage.Headers.Authorization =
+                new AuthenticationHeaderValue("Basic", Convert.ToBase64String(byteArray));
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs:61]: So is this the only difference between the two in the implementation?

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 2]
```diff
@@ -26,6 +32,45 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
...
+            requestMessage.Content = requestContent;
+            requestMessage.Version = Backchannel.DefaultRequestVersion;
+            var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs]:
```suggestion

            using var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
```

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 3]
```diff
@@ -46,5 +91,14 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
             await Options.Events.CreatingTicket(context);
             return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
         }
+
+        private static async Task<string> Display(HttpResponseMessage response)
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs]:
```suggestion
        private static async Task<string> DisplayAsync(HttpResponseMessage response)
```

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 4]
```diff
@@ -26,6 +32,45 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
...
+            requestMessage.Content = requestContent;
+            requestMessage.Version = Backchannel.DefaultRequestVersion;
+            var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
+            if (response.IsSuccessStatusCode)
+            {
+                var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+                return OAuthTokenResponse.Success(payload);
+            }
+
+            var error = "OAuth token endpoint failure: " + await Display(response);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs]:
```suggestion
            var error = "OAuth token endpoint failure: " + await DisplayAsync(response);
```

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 5]
```diff
@@ -26,6 +32,45 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
...
 
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(OAuthCodeExchangeContext context)
+        {
+            var tokenRequestParameters = new Dictionary<string, string>()
+            {
+                { "client_id", Options.ClientId },
+                { "redirect_uri", context.RedirectUri },
+                { "client_secret", Options.ClientSecret },
+                { "code", context.Code },
+                { "grant_type", "authorization_code" },
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs]:
```suggestion
                ["client_id"] = Options.ClientId,
                ["redirect_uri"] = context.RedirectUri,
                ["client_secret"] = Options.ClientSecret,
                ["code"] = context.Code,
                ["grant_type"] = "authorization_code",
```

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 6]
```diff
@@ -26,6 +32,45 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
...
+            };
+
+            // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+            if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier) &&
+                !string.IsNullOrEmpty(codeVerifier))
+            {
+                tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs]:
```suggestion
                tokenRequestParameters[OAuthConstants.CodeVerifierKey] = codeVerifier;
```

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 7]
```diff
@@ -26,6 +32,45 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
...
+                context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+            }
+
+            var requestContent = new FormUrlEncodedContent(tokenRequestParameters!);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs]:
```suggestion
            using var requestContent = new FormUrlEncodedContent(tokenRequestParameters!);
```

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 8]
```diff
@@ -26,6 +32,45 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
...
+
+            using var requestMessage = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+            requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            byte[] byteArray = Encoding.ASCII.GetBytes(Options.ClientId + ":" + Options.ClientSecret);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs:59]:
```suggestion

            byte[] byteArray = Encoding.ASCII.GetBytes(Options.ClientId + ":" + Options.ClientSecret);
```

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs [hunk 9]
```diff
@@ -46,5 +91,14 @@ public class NotionAuthenticationHandler : OAuthHandler<NotionAuthenticationOpti
             await Options.Events.CreatingTicket(context);
             return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
         }
+
+        private static async Task<string> Display(HttpResponseMessage response)
+        {
+            var output = new StringBuilder();
+            output.Append("Status: " + response.StatusCode + ";");
+            output.Append("Headers: " + response.Headers + ";");
+            output.Append("Body: " + await response.Content.ReadAsStringAsync() + ";");
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs]:
```suggestion
            output.Append("Status: ").Append(response.StatusCode).Append(';');
            output.Append("Headers: ").Append(response.Headers).Append(';');
            output.Append("Body: ").Append(await response.Content.ReadAsStringAsync()).Append(';');
```

File: src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs (hunk 1 above)
Comment (mifopen) [reply src/AspNet.Security.OAuth.Notion/NotionAuthenticationHandler.cs:61]: Yep

**Comment by martincostello** [comment]
Please verify the Notion provider works as expected using the [`5.0.7-preview.21308.75`](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.Notion/5.0.7-preview.21308.75) build from our MyGet feed. Once validated as working against the real service, I'll push the fix to NuGet.org.

**Comment by mifopen** [comment]
Successfully verified the version with the fix against the real Notion API

**Comment by martincostello** [comment]
The fix is now available from [NuGet.org](https://www.nuget.org/packages/AspNet.Security.OAuth.Notion/5.0.7) - thanks for the fix 🦸

**Pull request merged**